File: test/org/jgll/disambiguation/precedence/PrecedenceTest10.java
package org.jgll.disambiguation.precedence;

import static org.jgll.util.CollectionsUtil.*;
import static org.junit.Assert.*;

import org.jgll.grammar.Grammar;
import org.jgll.grammar.GrammarSlotRegistry;
import org.jgll.grammar.precedence.OperatorPrecedence;
import org.jgll.grammar.symbol.Character;
import org.jgll.grammar.symbol.Keyword;
import org.jgll.grammar.symbol.Nonterminal;
import org.jgll.grammar.symbol.Rule;
import org.jgll.parser.GLLParser;
import org.jgll.parser.ParseResult;
import org.jgll.parser.ParserFactory;
import org.jgll.sppf.IntermediateNode;
import org.jgll.sppf.NonterminalNode;
import org.jgll.sppf.PackedNode;
import org.jgll.sppf.SPPFNode;
import org.jgll.sppf.SPPFNodeFactory;
import org.jgll.sppf.TerminalNode;
import org.jgll.util.Input;
import org.junit.Before;
import org.junit.Test;

/**
 * 
 * E ::= E X    (none)
 * 	   > E ; E  (right)
 * 	   > - E
 *     | a
 * 
 * X ::= X , E
 *     | , E
 * 
 * @author Ali Afroozeh
 *
 */
public class PrecedenceTest10 {
	
	private Nonterminal E = Nonterminal.withName("E");
	private Nonterminal X = Nonterminal.withName("X");
	private Character a = Character.from('a');
	private Character comma = Character.from(',');
	private Character semicolon = Character.from(';');
	private Keyword min = Keyword.from("-");
	
	private GLLParser parser;
	private Grammar grammar;
	
	@Before
	public void init() {
		
		Grammar.Builder builder = new Grammar.Builder();
		
		// E ::= E X
		Rule rule1 = new Rule(E, list(E, X));
		builder.addRule(rule1);
		
		// E ::= E ; E
		Rule rule2 = new Rule(E, list(E, semicolon, E));
		builder.addRule(rule2);
		
		// E ::= - E
		Rule rule3 = new Rule(E, list(min, E));
		builder.addRule(rule3);
		
		// E ::= a
		Rule rule4 = new Rule(E, list(a));
		builder.addRule(rule4);
		
		// X ::= X , E
		Rule rule5 = new Rule(X, list(X, comma, E));
		builder.addRule(rule5);
		
		// X ::= , E
		Rule rule6 = new Rule(X, list(comma, E));
		builder.addRule(rule6);
		
		
		OperatorPrecedence operatorPrecedence = new OperatorPrecedence();

		// (E, .E X, E ";" E)
		operatorPrecedence.addPrecedencePattern(E, rule1, 0, rule2);
		
		// (E, E .X, E ";" E)
		operatorPrecedence.addPrecedencePattern(E, rule1, 1, rule2);
		
		// (E, .E X, - E)
		operatorPrecedence.addPrecedencePattern(E, rule1, 0, rule3);		
		
		// (E, .E ";" E, - E)
		operatorPrecedence.addPrecedencePattern(E, rule2, 0, rule3);
		
		grammar = operatorPrecedence.transform(builder.build());
	}
	
	@Test
	public void test1() {
		Input input = Input.fromString("a,-a;a");
		parser = ParserFactory.newParser(grammar, input);
		ParseResult result = parser.parse(input, grammar.toGrammarGraph(), "E");
		assertTrue(result.isParseSuccess());
		assertEquals(0, result.asParseSuccess().getParseStatistics().getCountAmbiguousNodes());
		assertTrue(result.asParseSuccess().getRoot().deepEquals(getSPPF(parser.getRegistry())));
	}
	
	private SPPFNode getSPPF(GrammarSlotRegistry registry) {
		SPPFNodeFactory factory = new SPPFNodeFactory(registry);
		NonterminalNode node1 = factory.createNonterminalNode("E", 0, 0, 6).init();
		PackedNode node2 = factory.createPackedNode("E ::= E1 X1 .", 1, node1);
		NonterminalNode node3 = factory.createNonterminalNode("E", 1, 0, 1).init();
		PackedNode node4 = factory.createPackedNode("E1 ::= a .", 0, node3);
		TerminalNode node5 = factory.createTerminalNode("a", 0, 1);
		node4.addChild(node5);
		node3.addChild(node4);
		NonterminalNode node6 = factory.createNonterminalNode("X", 1, 1, 6).init();
		PackedNode node7 = factory.createPackedNode("X1 ::= , E2 .", 2, node6);
		TerminalNode node8 = factory.createTerminalNode(",", 1, 1);
		NonterminalNode node9 = factory.createNonterminalNode("E", 2, 2, 6).init();
		PackedNode node10 = factory.createPackedNode("E2 ::= - E .", 3, node9);
		TerminalNode node11 = factory.createTerminalNode("-", 2, 1);
		NonterminalNode node12 = factory.createNonterminalNode("E", 0, 3, 6).init();
		PackedNode node13 = factory.createPackedNode("E ::= E3 ; E .", 5, node12);
		IntermediateNode node14 = factory.createIntermediateNode("E ::= E3 ; . E", 3, 5).init();
		PackedNode node15 = factory.createPackedNode("E ::= E3 ; . E", 4, node14);
		NonterminalNode node16 = factory.createNonterminalNode("E", 3, 3, 4).init();
		PackedNode node17 = factory.createPackedNode("E3 ::= a .", 3, node16);
		TerminalNode node18 = factory.createTerminalNode("a", 3, 1);
		node17.addChild(node18);
		node16.addChild(node17);
		TerminalNode node19 = factory.createTerminalNode(";", 4, 1);
		node15.addChild(node16);
		node15.addChild(node19);
		node14.addChild(node15);
		NonterminalNode node20 = factory.createNonterminalNode("E", 0, 5, 6).init();
		PackedNode node21 = factory.createPackedNode("E ::= a .", 5, node20);
		TerminalNode node22 = factory.createTerminalNode("a", 5, 1);
		node21.addChild(node22);
		node20.addChild(node21);
		node13.addChild(node14);
		node13.addChild(node20);
		node12.addChild(node13);
		node10.addChild(node11);
		node10.addChild(node12);
		node9.addChild(node10);
		node7.addChild(node8);
		node7.addChild(node9);
		node6.addChild(node7);
		node2.addChild(node3);
		node2.addChild(node6);
		node1.addChild(node2);
		return node1;
	}
	

}

File: test/org/iguana/disambiguation/precedence/PrecedenceTest10.scala
/*
 * Copyright (c) 2015, Ali Afroozeh and Anastasia Izmaylova, Centrum Wiskunde & Informatica (CWI)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this
 *    list of conditions and the following disclaimer in the documentation and/or
 *    other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
 * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 *
 */
package org.iguana.disambiguation.precedence

import iguana.utils.input.Input
import org.iguana.grammar.Grammar
import org.iguana.grammar.iggy.IggyParser
import org.iguana.grammar.patterns.PrecedencePattern
import org.iguana.grammar.precedence.OperatorPrecedence
import org.iguana.grammar.symbol.{Nonterminal, Rule, Terminal}
import org.iguana.parser.{Iguana, ParseResult}

/**
  *
  * E ::= E X    (none)
  * > E ; E  (right)
  * > - E
  * | a
  *
  * X ::= X , E
  * | , E
  *
  * @author Ali Afroozeh
  *
  */
class PrecedenceTest10 extends FunSuite {

  val E = Nonterminal.withName("E")
  val X = Nonterminal.withName("X")
  val a = Terminal.from(Character.from('a'))
  val comma = Terminal.from(Character.from(','))
  val semicolon = Terminal.from(Character.from(';'))
  val min = Terminal.from(Character.from('-'))

  val grammar = {
    val builder: Grammar.Builder = new Grammar.Builder
    val rule1: Rule = Rule.withHead(E).addSymbols(E, X).build
    builder.addRule(rule1)
    val rule2: Rule = Rule.withHead(E).addSymbols(E, semicolon, E).build
    builder.addRule(rule2)
    val rule3: Rule = Rule.withHead(E).addSymbols(min, E).build
    builder.addRule(rule3)
    val rule4: Rule = Rule.withHead(E).addSymbol(a).build
    builder.addRule(rule4)
    val rule5: Rule = Rule.withHead(X).addSymbols(X, comma, E).build
    builder.addRule(rule5)
    val rule6: Rule = Rule.withHead(X).addSymbols(comma, E).build
    builder.addRule(rule6)
    val list: java.util.List[PrecedencePattern] = new java.util.ArrayList[PrecedencePattern]
    list.add(PrecedencePattern.from(rule1, 0, rule2))
    list.add(PrecedencePattern.from(rule1, 1, rule2))
    list.add(PrecedencePattern.from(rule1, 0, rule3))
    list.add(PrecedencePattern.from(rule2, 0, rule3))
    val operatorPrecedence: OperatorPrecedence = new OperatorPrecedence(list)
    operatorPrecedence.transform(builder.build)
  }

  test("Grammar") {
    assert(grammar == expectedGrammar)
  }

  test("Parser") {
    val input: Input = Input.fromString("a,-a;a")
    val result: ParseResult = Iguana.parse(input, grammar, Nonterminal.withName("E"))
    assert(result.isParseSuccess)
    assert(result.asParseSuccess.getStatistics.getCountAmbiguousNodes == 0)
  }

  val expectedGrammar = {
    val s =
      """
        | E5 ::=   E1 X1
        |     |    E3 ';' E5
        |     |    'a'
        |
        | E  ::=   E1 X1
        |     |    E3 ';' E
        |     |    '-' E
        |     |    'a'
        |
        | X  ::=   X ',' E
        |     |    ',' E
        |
        | X1 ::=   X ',' E2
        |     |    ',' E2
        |
        | X2 ::=   X ',' E4
        |     |    ',' E4
        |
        | X3 ::=   X ',' E2
        |     |    ',' E2
        |
        | E1 ::=   E1 X2
        |     |    'a'
        |
        | X4 ::=   X ',' E4
        |     |    ',' E4
        |
        | E2 ::=   E1 X3
        |     |    '-' E
        |     |    'a'
        |
        | E3 ::=   E1 X4
        |     |    E3 ';' E5
        |     |    'a'
        |
        | E4 ::=   E1 X3
        |     |    'a'
        |
      """.stripMargin

    IggyParser.getGrammar(Input.fromString(s))
  }

}